### PR TITLE
Move java new side-effect removal into separate pass

### PR DESCRIPTION
**Pull request by peterschrammel**
removes java-specifc code from goto_convertt

### CODE REVIEW

File: src/goto-programs/goto_convert.cpp [hunk 1]
```diff
@@ -753,43 +753,15 @@ void goto_convertt::convert_assign(
...
-    code_assignt new_assign(code);
-    new_assign.lhs()=lhs;
-    new_assign.rhs()=rhs;
-
-    copy(new_assign, ASSIGN, dest);
-  }
-  else
-  {
-    clean_expr(rhs, dest);
-
```
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp]: I believe this is not preserved properly, and that's making the tests fail.

File: src/goto-programs/goto_convert.cpp [hunk 2]
```diff
@@ -753,43 +753,15 @@ void goto_convertt::convert_assign(
...
-          rhs.get(ID_statement)==ID_java_new_array)
-  {
-    Forall_operands(it, rhs)
-      clean_expr(*it, dest);
-
-    do_java_new_array(lhs, to_side_effect_expr(rhs), dest);
-  }
-  else if(
-    rhs.id() == ID_side_effect &&
-    (rhs.get(ID_statement) == ID_allocate ||
```
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp]: I don't think the `ID_allocate` case is preserved?

**Review by tautschnig — CHANGES_REQUESTED** [review]

**Review by smowton — APPROVED** [review]
Note this definitely needs concurrent test-gen change (@thk123)

**Comment by smowton** [comment]
Scratch that; https://github.com/diffblue/test-gen/pull/1757

**Comment by peterschrammel** [comment]
There's still a unit test failure that I have to fix.

**Comment by peterschrammel** [comment]
@tautschnig, CI is passing, please re-review.

**Review by tautschnig — APPROVED** [review]

**Pull request merged**